starlite/contrib/sqlalchemy/repository.py

```diff
@@ -103,17 +111,29 @@ async def add_many(self, data: list[ModelT]) -> list[ModelT]:
         Returns:
             The added instances.
         """
+        data_to_insert: list[dict[str, Any]] = [v.to_dict() if isinstance(v, self.model_type) else v for v in data]  # type: ignore
         with wrap_sqlalchemy_exception():
-            instances: list[ModelT] = await self._execute(
-                insert(
-                    self.model_type,
+            if self._supports_bulk_insert:
+                data: list[ModelT] = await self.session.execute(  # type: ignore
+                    insert(self.model_type).returning(self.model_type),
+                    data_to_insert,
                 )
-                .values([v.to_dict() if isinstance(v, self.model_type) else v for v in data])
-                .returning(self.model_type),  # type: ignore
+                for instance in data:
+                    self.session.expunge(instance)
+                return data
+            # when bulk insert with returning isn't supported, we ensure there is a unique ID on each record so that we can refresh the records after insert.
+            # ensure we have a UUID ID on each record so that we can refresh the data before returning
+            for datum in data_to_insert:
+                if datum.get("id", None) is None:
+                    datum.update({"id": uuid4()})
+            await self.session.execute(
+                insert(self.model_type),
+                data_to_insert,
+            )
+            # select the records we just updated.
+            return await self.list(
+                CollectionFilter(field_name="id", values=[datum.get("id") for datum in data_to_insert])
```

Comment:
I think it's better to store the IDs within the loop above, so you don't have to loop again here.